api/graphql/types/reservation/filtersets.py

```diff
@@ -145,22 +146,45 @@ def get_only_with_permission(self, qs: QuerySet, name: str, value: bool) -> Quer
         if not value:
             return qs
 
-        user = self.request.user
-        viewable_units = get_units_where_can_view_reservations(user)
+        user: AnyUser = self.request.user
         if user.is_anonymous:
             return qs.none()
-        return qs.filter(Q(reservation_unit__unit__in=viewable_units) | Q(user=user)).distinct()
+        if user.is_superuser:
+            return qs
+        if has_general_permission(user, GeneralPermissionChoices.CAN_VIEW_RESERVATIONS):
+            return qs
+
+        unit_permission = UnitPermissionChoices.CAN_VIEW_RESERVATIONS.value
+        unit_ids = [pk for pk, perms in user.unit_permissions.items() if unit_permission in perms]
+        unit_group_ids = [pk for pk, perms in user.unit_group_permissions.items() if unit_permission in perms]
+
+        return qs.filter(
+            Q(reservation_unit__unit__in=unit_ids)  #
+            | Q(reservation_unit__unit__unit_groups__in=unit_group_ids)
+        )
 
     def get_only_with_handling_permission(self, qs: QuerySet, name: str, value: bool) -> QuerySet:
         if not value:
             return qs
 
-        user = self.request.user
+        user: AnyUser = self.request.user
+
         if user.is_anonymous:
             return qs.none()
+        if user.is_superuser:
+            return qs
+        if has_general_permission(user, GeneralPermissionChoices.CAN_MANAGE_RESERVATIONS):
+            return qs
 
-        units = get_units_with_permission(user, permission="can_manage_reservations")
-        return qs.filter(Q(user=user) | Q(reservation_unit__unit__in=units)).distinct()
+        unit_permission = UnitPermissionChoices.CAN_MANAGE_RESERVATIONS.value
+        unit_ids = [pk for pk, perms in user.unit_permissions.items() if unit_permission in perms]
+        unit_group_ids = [pk for pk, perms in user.unit_group_permissions.items() if unit_permission in perms]
+
+        return qs.filter(
+            Q(user=user)
```

Comment:
Overlooked this earlier, let's remove the `user=user` condition from this filter as well (staff members cannot "handle" reservations created by themselves if no longer explicitly assigned permission on the unit, same as viewing them), other than that the PR good to go.
It was added in [TILA-3132](https://helsinkisolutionoffice.atlassian.net/browse/TILA-3132) but somehow ended up not being yet taken into use in the frontend (created [TILA-3351](https://helsinkisolutionoffice.atlassian.net/browse/TILA-3351) for that): https://github.com/City-of-Helsinki/tilavarauspalvelu-ui/blob/363c324abafb754aabbf7855e2ee18d1cb7a7b40/apps/admin-ui/src/common/queries.tsx#L132

Reply:
Ok, removed this and simplified permissions even further, removing some helpers that were only used once, using proper permission enums, simplifying queries, etc.